File: ProjetoMiniC/src/lexer/scanner.py
"""
scanner.py
==========
Mecanismo principal de análise léxica (Scanner). Converte o código fonte
em uma lista sequencial de tokens e identifica violações sintáticas/léxicas.
"""

from __future__ import annotations

import sys
from pathlib import Path
from typing import Dict, List, Optional, Union

DIRETORIO_ATUAL = Path(__file__).resolve().parent
if str(DIRETORIO_ATUAL) not in sys.path:
    sys.path.insert(0, str(DIRETORIO_ATUAL))

try:
    from .analysis_result import ResultadoAnalise
    from .errors import (
        ErroCaractereNaoTerminado,
        ErroCadeiaNaoTerminada,
        ErroComentarioNaoTerminado,
        ErroIdentificadorInvalido,
        ErroLexico,
        ErroLiteralRealMalformado,
        ErroSimboloInvalido,
    )
    from .jsonl_serializer import serialize_errors_jsonl, serialize_tokens_jsonl
    from .token_types import PALAVRAS_RESERVADAS, TokenType
    from .tokens import Token
except (ImportError, ValueError):
    from ProjetoMiniC.src.lexer.analysis_result import ResultadoAnalise
    from ProjetoMiniC.src.lexer.errors import (
        ErroCaractereNaoTerminado,
        ErroCadeiaNaoTerminada,
        ErroComentarioNaoTerminado,
        ErroIdentificadorInvalido,
        ErroLexico,
        ErroLiteralRealMalformado,
        ErroSimboloInvalido,
    )
    from ProjetoMiniC.src.lexer.jsonl_serializer import serialize_errors_jsonl, serialize_tokens_jsonl
    from ProjetoMiniC.src.lexer.token_types import PALAVRAS_RESERVADAS, TokenType
    from ProjetoMiniC.src.lexer.tokens import Token


class Scanner:
    """Analisador léxico por autômato finito determinístico de leitura direta."""

    # Mapeamento direto de pontuações de caractere único
    OPERADORES_SIMPLES: Dict[str, TokenType] = {
        "+": TokenType.PLUS,
        "-": TokenType.MINUS,
        "*": TokenType.STAR,
        "%": TokenType.PERCENT,
        ".": TokenType.DOT,
        "(": TokenType.LPAREN,
        ")": TokenType.RPAREN,
        "{": TokenType.LBRACE,
        "}": TokenType.RBRACE,
        "[": TokenType.LBRACKET,
        "]": TokenType.RBRACKET,
        ";": TokenType.SEMI,
        ",": TokenType.COMMA,
    }

    def __init__(self, codigo_fonte: str):
        self.codigo_fonte: str = codigo_fonte
        self.tamanho: int = len(codigo_fonte)
        self.posicao: int = 0
        self.linha: int = 1
        self.coluna: int = 1
        self.tokens: List[Token] = []
        self.errors: List[LexicalError] = []

    def _at_end(self) -> bool:
        return self.pos >= self.length

    @staticmethod
    def _is_identifier_start(ch: str) -> bool:
        return ch == "_" or (ch.isascii() and ch.isalpha())

    @staticmethod
    def _is_identifier_continue(ch: str) -> bool:
        return Scanner._is_identifier_start(ch) or (ch.isascii() and ch.isdigit())

    def _peek(self, offset: int = 0) -> str:
        idx = self.pos + offset
        return self.source[idx] if idx < self.length else "\0"

    def _advance(self) -> str:
        ch = self.source[self.pos]
        self.pos += 1
        if ch == "\n":
            self.line += 1
            self.column = 1
        else:
            self.coluna += 1
        return caractere

    def _compara_e_avanca(self, esperado: str) -> bool:
        """Avança o ponteiro apenas se o próximo caractere corresponder ao caractere esperado."""
        if self._espiar() == esperado:
            self._avancar()
            return True
        return False

    def _adicionar_token(self, tipo: TokenType, lexema: str, linha: int, coluna: int,
                           atributo: Optional[Union[int, float, str]] = None) -> None:
        """Instancia e adiciona um token reconhecido na coleção principal."""
        self.tokens.append(Token(tipo, lexema, linha, coluna, atributo))

    def scan_tokens(self) -> List[Token]:
        """Varre iterativamente a fonte até o fim e anexa o token EOF final."""
        while not self._esta_no_fim():
            self._ignorar_espacos_em_branco()
            if self._esta_no_fim():
                break
            self._processar_proximo_token()
        self.tokens.append(Token(TokenType.EOF, "", self.linha, self.coluna, None))
        return self.tokens

    def analisar(self) -> ResultadoAnalise:
        """Executa a análise e compila o relatório formal no objeto ResultadoAnalise."""
        self.scan_tokens()
        return AnalysisResult(tokens=self.tokens, errors=self.errors)

    def _skip_whitespace(self) -> None:
        while not self._at_end() and self._peek() in " \t\r\n":
            self._advance()

    def _scan_token(self) -> None:
        start_line, start_col = self.line, self.column
        ch = self._advance()

        if self._is_identifier_start(ch):
            self._identifier(start_line, start_col, ch)
        elif ch.isascii() and ch.isdigit():
            self._number(start_line, start_col, ch)
        elif ch == '"':
            self._string(start_line, start_col)
        elif ch == "'":
            self._char_literal(start_line, start_col)
        elif ch == "/" and self._peek() == "/":
            self._line_comment()
        elif ch == "/" and self._peek() == "*":
            self._block_comment(start_line, start_col)
        else:
            self._operator_or_error(ch, start_line, start_col)

    def _identifier(self, line: int, col: int, first_char: str) -> None:
        lexeme = first_char
        while not self._at_end() and self._is_identifier_continue(self._peek()):
            lexeme += self._advance()

        ttype = RESERVED_WORDS.get(lexeme, TokenType.ID)
        attribute = lexeme if ttype is TokenType.ID else None
        self._add_token(ttype, lexeme, line, col, attribute)

    def _number(self, line: int, col: int, first_digit: str) -> None:
        digits = first_digit
        while not self._at_end() and self._peek().isascii() and self._peek().isdigit():
            digits += self._advance()

        # Caso i06: Identificador iniciado por dígito
        if not self._at_end() and self._is_identifier_start(self._peek()):
            letters = ""
            letters_col = self.column
            while not self._at_end() and self._is_identifier_continue(self._peek()):
                letters += self._advance()

            self.errors.append(InvalidIdentifierError(digits + letters, line, col))
            self._add_token(TokenType.NUM_INT, digits, line, col, int(digits))
            self._add_token(TokenType.ID, letters, line, letters_col, letters)
            return

        # Número real ou malformado (Caso i05)
        if self._peek() == ".":
            if not (self._peek(1).isascii() and self._peek(1).isdigit()):
                dot_col = self.column
                self._advance()
                self.errors.append(MalformedRealLiteralError(digits + ".", line, col))
                self._add_token(TokenType.NUM_INT, digits, line, col, int(digits))
                self._add_token(TokenType.DOT, ".", line, dot_col, None)
                return

            lexeme = digits + self._advance()
            while not self._at_end() and self._peek().isascii() and self._peek().isdigit():
                lexeme += self._advance()

            self._adicionar_token(TokenType.NUM_FLOAT, lexema, linha, coluna, float(lexema))
            return

        self._adicionar_token(TokenType.NUM_INT, digitos, linha, coluna, int(digitos))

    def _string(self, line: int, col: int) -> None:
        start_pos = self.pos - 1
        content = ""
        closed = False

        while not self._esta_no_fim():
            if self._espiar() == "\n":
                break
            if self._espiar() == '"':
                self._avancar()
                fechado = True
                break
            ch = self._advance()
            if ch == "\\" and not self._at_end() and self._peek() != "\n":
                escape = self._advance()
                if escape in "nt\\\"'":
                    content += {"n": "\n", "t": "\t"}.get(escape, escape)
                    continue
                content += "\\" + escape
                continue
            content += ch

        if closed:
            lexeme = self.source[start_pos:self.pos]
            self._add_token(TokenType.STRING, lexeme, line, col, content)
        else:
            err_lexeme = self.source[start_pos:self.pos]
            self.errors.append(UnterminatedStringError(err_lexeme, line, col))

            # Mantém delimitadores finais no fluxo para recuperação dos fixtures.
            while self.pos > start_pos + 1 and self.source[self.pos - 1] in ") ;}]":
                self.pos -= 1
                self.column -= 1

    def _char_literal(self, line: int, col: int) -> None:
        if self._at_end() or self._peek() == "\n":
            self.errors.append(UnterminatedCharError("'", line, col))
            return

        ch = self._advance()
        raw = ch
        if ch == "\\":
            if self._at_end() or self._peek() == "\n":
                self.errors.append(UnterminatedCharError("'\\", line, col))
                return
            escape = self._advance()
            if escape not in "nt\\\"'":
                self.errors.append(UnterminatedCharError("'\\" + escape, line, col))
                return
            raw = "\\" + escape
            ch = {"n": "\n", "t": "\t"}.get(escape, escape)
        if self._match("'"):
            self._add_token(TokenType.CHAR_LITERAL, f"'{raw}'", line, col, ch)
        else:
            lexema = f"'{caractere}"
            self.erros.append(ErroCaractereNaoTerminado(lexema, linha, coluna))
            if self._espiar() == ";":
                self._avancar()

    def _processar_comentario_linha(self) -> None:
        """Descarta o restante da linha atual ao identificar comentários de linha (//)."""
        self._avancar()
        while not self._esta_no_fim() and self._espiar() != "\n":
            self._avancar()

    def _processar_comentario_bloco(self, linha: int, coluna: int) -> None:
        """Processa comentários de múltiplas linhas (/* ... */)."""
        posicao_inicial = self.posicao - 1
        self._avancar()
        while True:
            if self._esta_no_fim():
                lexema = self.codigo_fonte[posicao_inicial:]
                self.erros.append(ErroComentarioNaoTerminado(linha, coluna, lexema))
                return
            if self._espiar() == "*" and self._espiar(1) == "/":
                self._avancar()
                self._avancar()
                return
            self._avancar()

    def _processar_operador_ou_erro(self, caractere: str, linha: int, coluna: int) -> None:
        """Analisa operadores de um ou dois caracteres e aciona erro para símbolos desconhecidos."""
        if caractere == "=":
            if self._compara_e_avanca("="):
                self._adicionar_token(TokenType.EQ, "==", linha, coluna)
            else:
                self._adicionar_token(TokenType.ASSIGN, "=", linha, coluna)
        elif caractere == "!":
            if self._compara_e_avanca("="):
                self._adicionar_token(TokenType.NEQ, "!=", linha, coluna)
            else:
                self._adicionar_token(TokenType.NOT, "!", linha, coluna)
        elif caractere == "<":
            if self._compara_e_avanca("="):
                self._adicionar_token(TokenType.LE, "<=", linha, coluna)
            else:
                self._adicionar_token(TokenType.LT, "<", linha, coluna)
        elif caractere == ">":
            if self._compara_e_avanca("="):
                self._adicionar_token(TokenType.GE, ">=", linha, coluna)
            else:
                self._adicionar_token(TokenType.GT, ">", linha, coluna)
        elif caractere == "&":
            if self._compara_e_avanca("&"):
                self._adicionar_token(TokenType.AND, "&&", linha, coluna)
            else:
                self._sinalizar_invalido(caractere, linha, coluna)
        elif caractere == "|":
            if self._compara_e_avanca("|"):
                self._adicionar_token(TokenType.OR, "||", linha, coluna)
            else:
                self._sinalizar_invalido(caractere, linha, coluna)
        elif caractere == "/":
            self._adicionar_token(TokenType.SLASH, "/", linha, coluna)
        elif caractere in self.OPERADORES_SIMPLES:
            self._adicionar_token(self.OPERADORES_SIMPLES[caractere], caractere, linha, coluna)
        else:
            self._sinalizar_invalido(caractere, linha, coluna)

    def _sinalizar_invalido(self, caractere: str, linha: int, coluna: int) -> None:
        """Registra a presença de caracteres não reconhecidos."""
        self.erros.append(ErroSimboloInvalido(caractere, linha, coluna))

    def possui_erros(self) -> bool:
        """Informa se foram registrados erros na execução do scanner."""
        return len(self.erros) > 0

    def imprimir_tokens(self) -> None:
        """Exibe a lista dos tokens reconhecidos em formato de tabela no terminal."""
        cabecalho = f"{'TIPO':<14}{'LEXEMA':<26}{'LINHA':<7}{'COLUNA':<8}{'ATRIBUTO'}"
        print(cabecalho)
        print("-" * len(cabecalho))
        for token in self.tokens:
            nome, lexema, linha, coluna, atributo = token.para_linha_tabela()
            lexema_repr = repr(lexema)
            if len(lexema_repr) > 24:
                lexema_repr = lexema_repr[:21] + "...'"
            print(f"{nome:<14}{lexema_repr:<26}{linha:<7}{coluna:<8}{atributo}")

    def imprimir_erros(self) -> None:
        """Imprime os erros identificados de maneira legível."""
        if not self.erros:
            print("Nenhum erro léxico encontrado.")
            return
        print(f"{len(self.erros)} erro(s) léxico(s) encontrado(s):")
        for erro in self.erros:
            print(f"  [ERRO LÉXICO] {erro.diagnostico()}")


def main() -> int:
    """Função de entrada do CLI ao rodar diretamente o módulo scanner.py."""
    caminho_alvo: str | None = None
    modo_apenas_jsonl = True

    for arg in sys.argv[1:]:
        if arg == "--jsonl":
            modo_apenas_jsonl = True
        elif not arg.startswith("--") and caminho_alvo is None:
            caminho_alvo = arg

    if not caminho_alvo:
        print("Uso: python scanner.py <arquivo.c | arquivo.minic> [--jsonl]", file=sys.stderr)
        return 1

    arquivo = Path(caminho_alvo)
    if not arquivo.exists() or not arquivo.is_file():
        print(f"Erro: Arquivo '{caminho_alvo}' não encontrado.", file=sys.stderr)
        return 1

    try:
        conteudo = arquivo.read_text(encoding="utf-8")
    except OSError as err:
        print(f"Erro ao ler arquivo: {err}", file=sys.stderr)
        return 1

    scanner = Scanner(conteudo)
    scanner.scan_tokens()

    saida_tokens_json = serialize_tokens_jsonl(scanner.tokens)
    saida_erros_json = serialize_errors_jsonl(scanner.erros) if scanner.erros else ""

    if modo_apenas_jsonl:
        if saida_tokens_json:
            print(saida_tokens_json)
        if saida_erros_json:
            print(saida_erros_json, file=sys.stderr)
        return 2 if scanner.possui_erros() else 0

    print("=" * 80)
    print(f"Análise Léxica - Arquivo: {arquivo.name}")
    print("=" * 80)
    print("Tokens reconhecidos:")
    scanner.imprimir_tokens()
    print("-" * 80)
    print("Diagnóstico:")
    scanner.imprimir_erros()

    print("-" * 80)
    print("Saída JSONL (Tokens):")
    print(saida_tokens_json if saida_tokens_json else "(vazio)")

    if scanner.erros:
        print("-" * 80)
        print("Saída JSONL (Erros):")
        print(saida_erros_json if saida_erros_json else "(vazio)")

    return 2 if scanner.possui_erros() else 0


if __name__ == "__main__":
    sys.exit(main())

File: C/scanner.c
#include "scanner.h"
#include "token_types.h"
#include "errors.h"
#include "util.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <stdbool.h>

void scanner_init(Scanner *sc, const char *source) {
    sc->source = source;
    sc->length = strlen(source);
    sc->pos = 0;
    sc->line = 1;
    sc->column = 1;

    sc->tokens_len = 0;
    sc->tokens_cap = 16;
    sc->tokens = (Token *)xmalloc(sc->tokens_cap * sizeof(Token));

    sc->errors_len = 0;
    sc->errors_cap = 8;
    sc->errors = (LexicalError *)xmalloc(sc->errors_cap * sizeof(LexicalError));
}

void scanner_free(Scanner *sc) {
    if (sc->tokens) {
        for (size_t i = 0; i < sc->tokens_len; i++) {
            token_free(&sc->tokens[i]);
        }
        free(sc->tokens);
        sc->tokens = NULL;
    }
    if (sc->errors) {
        for (size_t i = 0; i < sc->errors_len; i++) {
            error_free(&sc->errors[i]);
        }
        free(sc->errors);
        sc->errors = NULL;
    }
}

static void add_token(Scanner *sc, TokenType type, const char *lexeme, int line, int col, const char *attr) {
    if (sc->tokens_len >= sc->tokens_cap) {
        sc->tokens_cap *= 2;
        sc->tokens = (Token *)xrealloc(sc->tokens, sc->tokens_cap * sizeof(Token));
    }
    sc->tokens[sc->tokens_len++] = token_create(type, lexeme, line, col, attr);
}

static void add_error(Scanner *sc, LexicalError err) {
    if (sc->errors_len >= sc->errors_cap) {
        sc->errors_cap *= 2;
        sc->errors = (LexicalError *)xrealloc(sc->errors, sc->errors_cap * sizeof(LexicalError));
    }
    sc->errors[sc->errors_len++] = err;
}

static bool at_end(const Scanner *s) {
    return s->pos >= s->length;
}

static char peek(const Scanner *s, int offset) {
    size_t idx = s->pos + offset;
    return (idx < s->length) ? s->source[idx] : '\0';
}

static char advance(Scanner *s) {
    char ch = s->source[s->pos++];
    if (ch == '\n') {
        s->line++;
        s->column = 1;
    } else {
        s->column++;
    }
    return ch;
}

static bool match(Scanner *s, char expected) {
    if (peek(s, 0) == expected) {
        advance(s);
        return true;
    }
    return false;
}

static void skip_whitespace(Scanner *s) {
    while (!at_end(s) && (peek(s, 0) == ' ' || peek(s, 0) == '\t' || peek(s, 0) == '\r' || peek(s, 0) == '\n')) {
        advance(s);
    }
}

static void scan_identifier(Scanner *s, int line, int col, char first_char) {
    char buffer[256];
    int len = 0;
    buffer[len++] = first_char;

    while (!at_end(s) && (isalnum((unsigned char)peek(s, 0)) || peek(s, 0) == '_')) {
        if (len < 255) buffer[len++] = advance(s);
        else advance(s);
    }
    buffer[len] = '\0';

    TokenType type = lookup_reserved_word(buffer);
    const char *attr = (type == ID) ? buffer : NULL;
    add_token(s, type, buffer, line, col, attr);
}

static void scan_number(Scanner *s, int line, int col, char first_digit) {
    char buffer[256];
    int len = 0;
    buffer[len++] = first_digit;

    while (!at_end(s) && isdigit((unsigned char)peek(s, 0))) {
        if (len < 255) buffer[len++] = advance(s);
        else advance(s);
    }
    buffer[len] = '\0';

    /* Identificador inválido iniciado por dígito */
    if (!at_end(s) && (isalpha((unsigned char)peek(s, 0)) || peek(s, 0) == '_')) {
        char letters[256];
        int l_len = 0;
        int l_col = s->column;
        while (!at_end(s) && (isalnum((unsigned char)peek(s, 0)) || peek(s, 0) == '_')) {
            if (l_len < 255) letters[l_len++] = advance(s);
            else advance(s);
        }
        letters[l_len] = '\0';

        char full_inv[512];
        snprintf(full_inv, sizeof(full_inv), "%s%s", buffer, letters);
        add_error(s, make_invalid_identifier_error(full_inv, line, col));
        add_token(s, NUM_INT, buffer, line, col, buffer);
        add_token(s, ID, letters, line, l_col, letters);
        return;
    }

    /* Número real malformado ou válido */
    if (peek(s, 0) == '.') {
        if (!isdigit((unsigned char)peek(s, 1))) {
            int dot_col = s->column;
            advance(s);
            char malformed[260];
            snprintf(malformed, sizeof(malformed), "%s.", buffer);
            add_error(s, make_malformed_real_literal_error(malformed, line, col));
            add_token(s, NUM_INT, buffer, line, col, buffer);
            add_token(s, DOT, ".", line, dot_col, NULL);
            return;
        }

        if (len < 255) buffer[len++] = advance(s);
        while (!at_end(s) && isdigit((unsigned char)peek(s, 0))) {
            if (len < 255) buffer[len++] = advance(s);
            else advance(s);
        }
        buffer[len] = '\0';
        add_token(s, NUM_FLOAT, buffer, line, col, buffer);
        return;
    }

    add_token(s, NUM_INT, buffer, line, col, buffer);
}

static void scan_string(Scanner *s, int line, int col) {
    char buffer[1024];
    int len = 0;
    bool closed = false;

    while (!at_end(s)) {
        if (peek(s, 0) == '\n') break;
        if (peek(s, 0) == '"') {
            advance(s);
            closed = true;
            break;
        }
        if (len < 1023) buffer[len++] = advance(s);
        else advance(s);
    }
    buffer[len] = '\0';

    if (closed) {
        char lexeme[1028];
        snprintf(lexeme, sizeof(lexeme), "\"%s\"", buffer);
        add_token(s, STRING, lexeme, line, col, buffer);
    } else {
        char err_lex[1028];
        snprintf(err_lex, sizeof(err_lex), "\"%s", buffer);
        add_error(s, make_unterminated_string_error(err_lex, line, col));
    }
}

static void scan_char(Scanner *s, int line, int col) {
    if (at_end(s) || peek(s, 0) == '\n') {
        add_error(s, make_unterminated_char_error("'", line, col));
        return;
    }

    char ch = advance(s);
    if (match(s, '\'')) {
        char lexeme[8];
        char attr[4] = {ch, '\0'};
        snprintf(lexeme, sizeof(lexeme), "'%c'", ch);
        add_token(s, CHAR_LITERAL, lexeme, line, col, attr);
    } else {
        char lexeme[8];
        snprintf(lexeme, sizeof(lexeme), "'%c", ch);
        add_error(s, make_unterminated_char_error(lexeme, line, col));
    }
}

static void scan_token_item(Scanner *s) {
    int start_line = s->line;
    int start_col = s->column;
    char ch = advance(s);

    if (isalpha((unsigned char)ch) || ch == '_') {
        scan_identifier(s, start_line, start_col, ch);
    } else if (isdigit((unsigned char)ch)) {
        scan_number(s, start_line, start_col, ch);
    } else if (ch == '"') {
        scan_string(s, start_line, start_col);
    } else if (ch == '\'') {
        scan_char(s, start_line, start_col);
    } else if (ch == '/' && peek(s, 0) == '/') {
        advance(s);
        while (!at_end(s) && peek(s, 0) != '\n') advance(s);
    } else if (ch == '/' && peek(s, 0) == '*') {
        advance(s);
        int start_pos = (int)s->pos - 2;
        while (true) {
            if (at_end(s)) {
                size_t comm_len = s->pos - start_pos;
                char *comm_buf = (char *)xmalloc(comm_len + 1);
                strncpy(comm_buf, &s->source[start_pos], comm_len);
                comm_buf[comm_len] = '\0';

                add_error(s, make_unterminated_comment_error(comm_buf, start_line, start_col));
                free(comm_buf);
                return;
            }
            if (peek(s, 0) == '*' && peek(s, 1) == '/') {
                advance(s);
                advance(s);
                return;
            }
            advance(s);
        }
    } else {
        switch (ch) {
            case '+': add_token(s, PLUS, "+", start_line, start_col, NULL); break;
            case '-': add_token(s, MINUS, "-", start_line, start_col, NULL); break;
            case '*': add_token(s, STAR, "*", start_line, start_col, NULL); break;
            case '/': add_token(s, SLASH, "/", start_line, start_col, NULL); break;
            case '%': add_token(s, PERCENT, "%", start_line, start_col, NULL); break;
            case '.': add_token(s, DOT, ".", start_line, start_col, NULL); break;
            case ';': add_token(s, SEMI, ";", start_line, start_col, NULL); break;
            case ',': add_token(s, COMMA, ",", start_line, start_col, NULL); break;
            case '(': add_token(s, LPAREN, "(", start_line, start_col, NULL); break;
            case ')': add_token(s, RPAREN, ")", start_line, start_col, NULL); break;
            case '{': add_token(s, LBRACE, "{", start_line, start_col, NULL); break;
            case '}': add_token(s, RBRACE, "}", start_line, start_col, NULL); break;
            case '[': add_token(s, LBRACKET, "[", start_line, start_col, NULL); break;
            case ']': add_token(s, RBRACKET, "]", start_line, start_col, NULL); break;
            case '=':
                if (match(s, '=')) add_token(s, EQ, "==", start_line, start_col, NULL);
                else add_token(s, ASSIGN, "=", start_line, start_col, NULL);
                break;
            case '!':
                if (match(s, '=')) add_token(s, NEQ, "!=", start_line, start_col, NULL);
                else add_token(s, NOT, "!", start_line, start_col, NULL);
                break;
            case '<':
                if (match(s, '=')) add_token(s, LE, "<=", start_line, start_col, NULL);
                else add_token(s, LT, "<", start_line, start_col, NULL);
                break;
            case '>':
                if (match(s, '=')) add_token(s, GE, ">=", start_line, start_col, NULL);
                else add_token(s, GT, ">", start_line, start_col, NULL);
                break;
            case '&':
                if (match(s, '&')) add_token(s, AND, "&&", start_line, start_col, NULL);
                else {
                    char str[2] = {ch, '\0'};
                    add_error(s, make_invalid_symbol_error(str, start_line, start_col));
                }
                break;
            case '|':
                if (match(s, '|')) add_token(s, OR, "||", start_line, start_col, NULL);
                else {
                    char str[2] = {ch, '\0'};
                    add_error(s, make_invalid_symbol_error(str, start_line, start_col));
                }
                break;
            default: {
                char str[2] = {ch, '\0'};
                add_error(s, make_invalid_symbol_error(str, start_line, start_col));
                break;
            }
        }
    }
}

void scanner_scan_tokens(Scanner *s) {
    while (!at_end(s)) {
        skip_whitespace(s);
        if (at_end(s)) break;
        scan_token_item(s);
    }
    add_token(s, TOK_EOF, "", s->line, s->column, NULL);
}

bool scanner_has_errors(const Scanner *s) {
    return s->errors_len > 0;
}

void scanner_print_tokens(const Scanner *s) {
    printf("%-18s%-26s%-7s%-8s%s\n", "TIPO", "LEXEMA", "LINHA", "COLUNA", "ATRIBUTO");
    printf("---------------------------------------------------------------------------\n");
    for (size_t i = 0; i < s->tokens_len; i++) {
        const Token *t = &s->tokens[i];
        char repr[64];
        snprintf(repr, sizeof(repr), "'%s'", t->lexeme);
        if (strlen(repr) > 24) {
            repr[21] = '.';
            repr[22] = '.';
            repr[23] = '.';
            repr[24] = '\'';
            repr[25] = '\0';
        }
        printf("%-18s%-26s%-7d%-8d%s\n",
               token_type_name(t->type),
               repr,
               t->line,
               t->column,
               t->attribute ? t->attribute : "");
    }
}

void scanner_print_errors(const Scanner *s) {
    if (s->errors_len == 0) {
        printf("Nenhum erro léxico encontrado.\n");
        return;
    }
    printf("%zu erro(s) léxico(s) encontrado(s):\n", s->errors_len);
    for (size_t i = 0; i < s->errors_len; i++) {
        char *diag = error_diagnostic(&s->errors[i]);
        printf("  [ERRO LÉXICO] %s\n", diag);
        free(diag);
    }
}